`addFiled` with `PRIMARY_KEY` doesn't apply `REQUIRED` modifier implicitly  (Original title - How to migrate a field's type the right way?)

Hey guys. I'm receiving an exception while trying to change a field's type.

Stacktrace: https://gist.github.com/Saketme/4b8f6274a0103165c79d
RealmObject: https://gist.github.com/Saketme/37894ccc5a1051cd8111
Migration: https://gist.github.com/Saketme/9ff0591dfdc45018e90b

In brief, I wanted to change a field's type from `long` to `String`. Retaining old values was not an option so the whole table had to be dropped. I however could not find a way to clear all items while migrating. Am I supposed to use `Realm.clear()` here?

That seemed to work, but I received a new exception:

`Caused by: io.realm.exceptions.RealmMigrationNeededException: Field 'id' does support null values in the existing Realm file. Remove @Required or @PrimaryKey from field 'id' or migrate using io.realm.internal.Table.convertColumnToNotNullable().`

The id field was previously a `long` field that cannot store null. Does Realm internally store primitive types as Objects?

--- a/changelog.txt
+++ b/changelog.txt
@@ -1,3 +1,6 @@
+0.87.2
+ * Fixed a bug when RealmObjectSchema.addField called with PRIMARY_KEY modifier, the field was not set as a required field (#2001).
+
 0.87.1
  * Upgraded to NDK R10e. Using gcc 4.9 for all architectures.
  * Updated Realm Core to 0.95.6

--- a/realm/realm-annotations/src/main/java/io/realm/annotations/PrimaryKey.java
+++ b/realm/realm-annotations/src/main/java/io/realm/annotations/PrimaryKey.java
@@ -26,7 +26,7 @@ import java.lang.annotation.Target;
  * should uniquely identify the object. Trying to insert an object with an existing primary key
  * will result in an {@link io.realm.exceptions.RealmException}.
  *
- * Primary keys on Strings also counts as having the {@link io.realm.annotations.Index} annotation.
+ * Primary keys also count as having the {@link Index} and {@link Required} annotations.
  *
  * Only one field pr. RealmObject class can have this annotation and it is only allowed on the following
  * types: String, short, int, long, byte

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmObjectSchemaTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmObjectSchemaTests.java
@@ -294,6 +294,7 @@ public class RealmObjectSchemaTests extends AndroidTestCase {
             String fieldName = "foo";
             schema.addField(fieldName, fieldType.getType(), FieldAttribute.PRIMARY_KEY);
             assertTrue(schema.hasPrimaryKey());
+            assertFalse(schema.isNullable(fieldName));
             schema.removeField(fieldName);
         }
     }

--- a/realm/realm-library/src/main/java/io/realm/FieldAttribute.java
+++ b/realm/realm-library/src/main/java/io/realm/FieldAttribute.java
@@ -30,7 +30,7 @@ public enum FieldAttribute {
     INDEXED,
 
     /**
-     * Mark a field as a primary key. This also implicitly mark it as {@link #INDEXED}.
+     * Mark a field as a primary key. This also implicitly mark it as {@link #INDEXED} and {@link #REQUIRED}.
      *
      * @see io.realm.annotations.PrimaryKey
      */

--- a/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
@@ -140,7 +140,11 @@ public final class RealmObjectSchema {
         }
 
         checkNewFieldName(fieldName);
-        boolean nullable = metadata.defaultNullable && !containsAttribute(attributes, FieldAttribute.REQUIRED);
+        boolean nullable = metadata.defaultNullable;
+        if (containsAttribute(attributes, FieldAttribute.REQUIRED) ||
+                containsAttribute(attributes, FieldAttribute.PRIMARY_KEY)) {
+            nullable = false;
+        }
         long columnIndex = table.addColumn(metadata.realmType, fieldName, nullable);
 
         try {

